### DISCUSSION TITLE
SQL error: Custom table that references org.opencrx.security.realm1.Principal

### DISCUSSION THREAD

**Discussion by nyemiks** [opencrx/opencrx#144] [Q&A] [answered]
I created a custom opencrx project so that I extend opencrx models. I modelled a new entity called "RefreshToken". The purpose of this table is to persist jwt refresh tokens for use in my spring mvc app that interacts with the opencrx back end. The RefreshToken has a non association relationship with org.opencrx.security.realm1.Principal class. The goal of modelling this relationship is so that I can perform a query that fetches all refresh tokens that belong to a given Principal.

  For example,

	   			 RefreshTokenQuery refreshTokenQuery = (RefreshTokenQuery)pm.newQuery(RefreshToken.class);
	                refreshTokenQuery.thereExistsUserPrincipal().name().equalTo(principalId);

	                 SegmentUtil segmentUtil = new SegmentUtil(providerName, segmentName);

	                com.spacio.crm.security1.jmi1.Segment segment = segmentUtil.getSecurityExtensionSegment(pm);

                        logger.info("query: " + refreshTokenQuery);
	                
	                List<RefreshToken> refreshTokenList = segment.getRefreshToken(refreshTokenQuery);
	                
	              The above should give me a list of tokens that belong to that principal.
	          
I have registered the entity in my kernel.properties

# refresh token
PERSISTENCE/type[780]=xri://@openmdx*com:spacio:crm:security1/provider/:*/segment/:*/refreshToken/:*
PERSISTENCE/typeName[780]=refreshToken
PERSISTENCE/dbObject[780]=OOCKE1_REFRESH_TOKEN
PERSISTENCE/dbObject2[780]=OOCKE1_REFRESH_TOKEN_
PERSISTENCE/pathNormalizeLevel[780]=(java.lang.Integer)2
PERSISTENCE/dbObjectFormat[780]=slicedWithIdAsKey

RefreshToken.openmdxjdo

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE openmdxjdo PUBLIC "-//openMDX//DTD Java Data Objects Metadata Extension 2.0//EN" "http://www.openmdx.org/dtd/openmdx-jdo_2_0.dtd">
<openmdxjdo>
  <package name="com.spacio.crm.security1.jpa3">    
	    <class name="RefreshToken" table="OOCKE1_REFRESH_TOKEN">
			<field name="openmdxjdoIdentity" visibility="none">
				<column name="OBJECT_ID"/>
			</field>
			<field name="segment" visibility="none">
				<column name="P$$PARENT"/>
			</field>            
            <field name="userPrincipal" visibility="cci">
			    <column name="PRINCIPAL"/>
		    </field>			
		</class>	
  </package>
</openmdxjdo>

When running the application, openmdx generates an incorrect sql error which eventually fails.

WARNING: Error when executing SQL statement
org.openmdx.base.exception.ServiceException: org.openmdx.base.exception.ServiceException: DefaultDomain.MEDIA_ACCESS_FAILURE
    BasicException[0]
	Class = org.openmdx.base.dataprovider.layer.persistence.jdbc.RestInteraction
	Method = retrieveDatabaseContent
	Line = 791
	ExceptionClass = org.openmdx.base.exception.ServiceException
	ExceptionDomain = DefaultDomain
	ExceptionCode = MEDIA_ACCESS_FAILURE
	Description = Error when executing SQL statement
	Timestamp = 2024-11-12T16:25:11.688Z
	Parameters:
	    0:	xri = xri://@openmdx*com.spacio.crm.security1/provider/CRX/segment/Oak/refreshToken
	    1:	statement = SELECT  v.*, v.dtype AS p$$dtype, v.user_principal AS p$$user_principal FROM OOCKE1_REFRESH_TOKEN v WHERE (1=1) AND ((v.object_id > ?) AND (v.object_id < ?)) AND (((v.dtype IN (?)))) AND (((v.user_principal IN (SELECT object_id FROM null vv WHERE (1=1) AND (((vv.name IN (?)))) AND (((vv.dtype IN (?, ?)))))))) ORDER BY v.object_id
	    2:	parameters = [refreshToken/CRX/Oak/, refreshToken/CRX/Oak0, com:spacio:crm:security1:RefreshToken, taka1, org:opencrx:security:realm1:Principal, com:spacio:crm:security1:Principal]
	    3:	sqlErrorCode = 156
	    4:	sqlState = S0001
	StackTrace:
	    at org.openmdx.base.dataprovider.layer.persistence.jdbc.RestInteraction.retrieveDatabaseContent(RestInteraction.java:791)
	    at org.openmdx.base.dataprovider.layer.persistence.jdbc.RestInteraction.find(RestInteraction.java:240)
	    at org.opencrx.kernel.layer.persistence.Media_2$RestInteraction.find(Media_2.java:341)
	    at org.opencrx.kernel.layer.persistence.Audit_2$RestInteraction.find(Audit_2.java:474)
	    at org.openmdx.base.rest.spi.AbstractRestInteraction.handleQuery(AbstractRestInteraction.java:484)
	    at org.openmdx.base.rest.spi.AbstractRestInteraction.execute(AbstractRestInteraction.java:361)
	

After the analysing the sql query, I believe the error is here 
"SELECT object_id FROM null vv WHERE (1=1) AND (((vv.name IN (?)))) AND (((vv.dtype IN (?, ?)))))))"

Why am I having a null value ?

I believe the correct sql should

SELECT object_id FROM OOMSE2_PRINCIPAL vv WHERE (1=1) AND (((vv.name IN (?)))) AND (((vv.dtype IN (?, ?)))))))

since "OOMSE2_PRINCIPAL" is the table that is being referenced.

Any ideas why am I having this error ?

Note: I have succesfully implemented model extensions to opencrx classes before. Though this is the first time I am referencing the Principal class.

I have attached screen shots of my custom model.

![image](https://github.com/user-attachments/assets/3a9d0978-a81f-452a-82a2-b0028f0ed800)

![image](https://github.com/user-attachments/assets/283a710d-7e84-4c37-b37d-c7e33973f6d1)

![image](https://github.com/user-attachments/assets/802d6ebb-273c-4b0f-b50c-6fcdf0b034b9)

**Comment by brunoamacos** [comment]
You need another mapping for principal within your model authority
 type=xri://@openmdx*com.spacio.crm.security1/provider/:*/segment/:*/realm/:*/principal/:*
 typeName unique name
 rest  as xri://@openmdx*org.openmdx.security.realm1/provider/:*/segment/:*/realm/:*/principal/:*

regards Bruno

**Comment by nyemiks** [comment, marked as answer]
I have fixed the problem. Rather I have a work around which I think is the proper solution. Apparently it had nothing to do with modelling.

The issue was related to the code I wrote.

  
			org.opencrx.security.realm1.jmi1.Principal principal =  getPrincipal(principalId);   // correct code. (fetch the principal object)
			  
     		 
	   			 RefreshTokenQuery refreshTokenQuery = (RefreshTokenQuery)pm.newQuery(RefreshToken.class);

                             //   refreshTokenQuery.thereExistsUserPrincipal().name().equalTo(principalId);   // previous code does not work

	                        refreshTokenQuery.thereExistsUserPrincipal().equalTo(principal);    // correct code.

                                SegmentUtil segmentUtil = new SegmentUtil(providerName, segmentName);

                com.spacio.crm.security1.jmi1.Segment segment = segmentUtil.getSecurityExtensionSegment(pm);

                    logger.info("query: " + refreshTokenQuery);
                
                List<RefreshToken> refreshTokenList = segment.getRefreshToken(refreshTokenQuery);
	       

          The above code works.
